plugins/rule_engines/irods_rule_language/src/configuration.cpp

```diff
@@ -420,20 +422,60 @@ int hash_rules(const std::vector<std::string> &irbs, const int pid, std::string
 
 class make_copy {
 public:
-        make_copy(const std::vector<std::string>& _irbs, const int _pid) : irbs_(_irbs), pid_(_pid) {
-            for(auto const &irb : _irbs) {
-                boost::filesystem::copy_file(get_rule_base_path(irb), get_rule_base_path_copy(irb, _pid));
-            }
+  make_copy(const std::vector<std::string>& _irbs)
+      : irbs_(_irbs)
+  {
+      for (auto const& irb : _irbs) {
+          std::string rulebase;
+          std::ifstream ifs(get_rule_base_path(irb));
+          rule_bases.insert(
+              {irb, std::string({std::istreambuf_iterator<char>(ifs), std::istreambuf_iterator<char>()})});
+      }
+  }
+
+  std::string& get_rulebase(const std::string& irb)
+  {
+      return rule_bases.at(irb);
+  }
+
+private:
+  std::unordered_map<std::string, std::string> rule_bases;
+  const std::vector<std::string> irbs_;
+};
+
+int hash_rules_with_copy(const std::vector<std::string>& irbs, std::string& digest, make_copy& copy)
+{
+    int status;
+    irods::Hasher hasher;
+    irods::error ret = irods::getHasher("md5", hasher);
+    if (!ret.ok()) {
+        status = ret.code();
+        rodsLogError(LOG_ERROR, status, "hash_rules_with_copy: cannot get hasher, status = %d", status);
+        return status;
+    }
+
+    for (auto const& irb : irbs) {
+        std::stringstream in_file(copy.get_rulebase(irb), std::ios::in | std::ios::binary);
+
+        std::string buffer_read;
+        buffer_read.resize(HASH_BUF_SZ);
+
+        while (in_file.read(&buffer_read[0], HASH_BUF_SZ)) {
+            hasher.update(buffer_read);
         }
-        ~make_copy() {
-            for(auto const &irb : irbs_) {
-                boost::filesystem::remove(get_rule_base_path_copy(irb, pid_));
+
+        if (in_file.eof()) {
+            if (in_file.gcount() > 0) {
+                buffer_read.resize(in_file.gcount());
+                hasher.update(buffer_read);
             }
         }
```

Comment:
If the EOF isn't reached, what does that mean for the rulebase?
Does anything care about the hash of these files?
Perhaps we should log a warning?

Reply:
Personally, it's hard to imagine a case where a `stringstream` doesn't hit EOF-- maybe it means the reads were interrupted above, somehow? I feel like the program's probably in a bad state if it can't read from something stored in memory...Could log a warning for that.